src/OpenSpaceToolkit/Physics/Environment/Atmospheric/Earth/NRLMSISE00.cpp

```diff
@@ -0,0 +1,291 @@
+{
+namespace environment
+{
+namespace atmospheric
+{
+namespace earth
+{
+
+using ostk::core::types::Integer;
+using ostk::core::types::Real;
+using ostk::core::types::Size;
+using ostk::core::ctnr::Array;
+
+using ostk::physics::coord::Frame;
+using ostk::physics::time::Scale;
+using ostk::physics::time::Instant;
+using ostk::physics::time::DateTime;
+using ostk::physics::time::Date;
+using ostk::physics::time::Time;
+
+using EarthCelestialBody = ostk::physics::env::obj::celest::Earth;
+using EarthGravitationalModel = ostk::physics::environment::gravitational::Earth;
+using ostk::physics::environment::atmospheric::earth::Manager;
+
+NRLMSISE00::NRLMSISE00()
+    : Model()
+{
+}
+
+NRLMSISE00* NRLMSISE00::clone() const
+{
+    return new NRLMSISE00(*this);
+}
+
+bool NRLMSISE00::isDefined() const
+{
+    return true;
+}
+
+void NRLMSISE00::computeApArray(double* outputArray, const Instant& anInstant) const
+{   
+    // Algorithm:
+    //
+    // Ap readings are given on a daily basis.
+    // We first create a timestamp in each day for which we will need data.
+    // Sometimes "now" and "now-9hr" land on the same day.
+    //
+    //  now-56hr                 now
+    //     v                      v
+    //     *        *        *    *
+    //     |--------|--------|----|
+    //        24hr     24hr   9hr
+    //
+    // Then, we fetch the Ap data for each of those days
+    // For each day, we get 8 3-hourly readings, starting at 00:00 UTC.
+    // They usually do not line up with the 24-hour periods.
+    //
+    //     *        *        *    *
+    //     |--------|--------|----|
+    //|........|........|........|........|
+    //   day1     day2    day3      day4
+    //
+    // We stack the data into a single array then find which data point 
+    // on day 1 lines up with the "now-56hr" timestamp
+    //
+    //     *        *        *    *
+    //     |--------|--------|----|
+    //|........|........|........|........|
+    //     ^
+    //  this one
+    // 
+    // Then use that as a starting index to find the correct data points and averages.
+
+    const Manager& spaceWeatherManager = Manager::Get();
+
+    // Fetch AP parameters for each day up to 57 hours ago
+    const Instant instant57HrPrevious = anInstant - Duration::Hours(57);
+    const Instant instant33HrPrevious = anInstant - Duration::Hours(33);
+    const Instant instant9HrPrevious = anInstant - Duration::Hours(9);
+
+    // Fetch daily AP array for each day and stack end-to-end
+    Array<Instant> fetchDays = {instant57HrPrevious, instant33HrPrevious, instant9HrPrevious};
+
+    // add today if not already covered by instant9HrPrevious
+    if (instant9HrPrevious.getDateTime(Scale::UTC).getDate().getDay() !=
+        anInstant.getDateTime(Scale::UTC).getDate().getDay())
+    {
+        fetchDays.add(anInstant);
+    }
+
+    // Fetch AP indices for all days. Stack them into one continuous array.
+    Array<Integer> apMultiDayArray = Array<Integer>::Empty();
```

Comment:
Nit:
Worth doing a `reserve` since you know the size here.
```suggestion
    Array<Integer> apMultiDayArray = Array<Integer>::Empty();
    apMultiDayArray.reserve(fetchDays.getSize());
```